### ISSUE TITLE
Recursive topk

### ISSUE THREAD

**Issue by crusaderky** [dask/dask#3395]
I have an array with shape (500000, ) and chunks=2500.
I need to pick the top 5000 elements out of it.

The current topk implementation is very bad for it, as:
1. it applies one kernel per chunk of the original input
2. concatenates all the outputs of the first pass together into a single chunk
3. performs a single, final sort

As my k is greater than the chunk size, effectively step 2 would create a single-chunk array or 500000 points.
The workaround would be to rechunk the array ahead of topk to larger chunks - which is often not the easiest thing to do when you're implementing a general-purpose library call.

As a solution I suggest to add a recursive=False option to both the method and the function; see prototype below:

https://gist.github.com/crusaderky/35d355d8124141aeea82213e0c7bccde

**Comment by crusaderky** [comment]
Updated gist with a better algorithm, which removes an unneeded sort() in every chunk

**Comment by jcrist** [comment]
The use case makes sense to me. Note that this is similar to the `split_every` option found on other dask array reductions (e.g. `mean`, main implementation in `dask.array.reductions.reduction`). I suggest if possible we follow the same api. Note that you won't be able to reuse the implementation in `dask.array.reductions.reduction` due to some assumptions it makes that aren't valid here, but a one-off implementation for `topk` shouldn't be that bad.

**Comment by crusaderky** [comment]
@jcrist uhm. wow. I now feel I tried reinventing the wheel (poorly).
Reimplemented everything from scratch with a lot of new features:

- now based on reduction(). The only hack I had to do was change the final chunk from (1, ) to (k, )
- standard support for recursively aggregating every n chunks, as with all other reductions
- support for multiple dimensions
- option to pick bottom k elements
- swapped a and k arguments (https://github.com/dask/dask/issues/2965)

https://gist.github.com/crusaderky/11948b5da1d5b2f0d5943dedca384e02

Proceeding to integrate and submitting a PR...
